cpp/oneapi/dal/table/backend/homogen_table_impl.cpp

```diff
@@ -23,250 +23,288 @@ namespace oneapi::dal::backend {
+            block.reset(data_, col_data + rows.start_idx * column_count, range_count);
+        }
     }
     else {
-        if (block.get_count() < range_count) {
-            block.reset(range_count);
+        if (block.get_count() < range_count || block.has_mutable_data() == false ||
+            has_array_data_kind(policy, block, kind) == false) {
+            reset_array(policy, block, range_count, kind);
         }
 
-        auto src_ptr = data_.get_data() + detail::get_data_type_size(feature_type) *
-                                              (idx + rows.start_idx * column_count);
-        backend::convert_vector(src_ptr,
+        auto src_ptr = data_.get_data() + detail::get_data_type_size(table_dtype) *
+                                              (column_index + rows.start_idx * column_count);
+        backend::convert_vector(policy,
+                                src_ptr,
                                 block.get_mutable_data(),
-                                feature_type,
+                                table_dtype,
                                 block_dtype,
-                                detail::get_data_type_size(feature_type) * column_count,
-                                sizeof(T),
+                                detail::get_data_type_size(table_dtype) * column_count,
+                                sizeof(Data),
                                 range_count);
     }
 }
 
-template <typename T>
-void homogen_table_impl::push_column(const array<T>& block, int64_t idx, const range& rows) {
+template <typename Policy, typename Data>
+void homogen_table_impl::push_column_impl(const Policy& policy,
+                                          const array<Data>& block,
+                                          int64_t column_index,
+                                          const range& rows) {
     // TODO: check inputs
 
     const int64_t row_count = get_row_count();
     const int64_t column_count = get_column_count();
     const int64_t range_count = rows.get_element_count(row_count);
-    const data_type block_dtype = detail::make_data_type<T>();
+    const data_type block_dtype = detail::make_data_type<Data>();
 
-    auto feature_type = meta_.get_data_type(0);
+    auto table_dtype = meta_.get_data_type(0);
     const int64_t row_offset =
-        detail::get_data_type_size(feature_type) * (idx + rows.start_idx * column_count);
+        detail::get_data_type_size(table_dtype) * (column_index + rows.start_idx * column_count);
 
-    if (block_dtype == feature_type && column_count == 1) {
+    if (block_dtype == table_dtype && column_count == 1) {
         if (reinterpret_cast<const void*>(data_.get_data() + row_offset) !=
             reinterpret_cast<const void*>(block.get_data())) {
-            data_.need_mutable_data();
+            make_mutable_data(policy, data_);
             auto dst_ptr = data_.get_mutable_data() + row_offset;
```

Comment:
It doesn't look consistent with check at line 249: data_.get_data() + row_offset against data_.get_mutable_data() + row_offset

Reply:
If array has mutable data, they both return the same pointer. Otherwise `get_mutable_data()` throws an exception.
Why it's inconsistent? Condition at the 249 line checks is block use the memory directly from the table. If not, we shall copy the data from the block to the table.